src/hotspot/share/oops/instanceKlass.cpp

```diff
@@ -4220,6 +4220,23 @@ bool InstanceKlass::should_clean_previous_versions_and_reset() {
   return ret;
 }
 
+// This nulls out jmethodIDs for all methods in 'klass'
+// It needs to be called explicitly for all previous versions of a class because these may not be cleaned up
+// during class unloading.
+// We can not use the jmethodID cache associated with klass directly because the 'previous' versions
+// do not have the jmethodID cache filled in. Instead, we need to lookup jmethodID for each method and this
+// is expensive - O(n) for one jmethodID lookup. For all contained methods it is O(n^2).
+// The reason for expensive jmethodID lookup for each method is that there is no direct link between method and jmethodID.
+void InstanceKlass::clear_jmethod_ids(InstanceKlass* klass) {
+  Array<Method*>* method_refs = klass->methods();
+  for (int k = 0; k < method_refs->length(); k++) {
+    Method* method = method_refs->at(k);
+    if (method != nullptr) {
+      method->clear_jmethod_id();
+    }
```

Comment:
This loops through the methods in the InstanceKlass that was a previous version klass, and clears the jmethodIDs for all the methods.  Will it clear the jmethodIDs for the EMCP methods also, and should it?
The jmethodID for EMCP methods are replaced with a the new version, so the Method* in this list won't find a matching jmethodID.  Maybe this can be restricted to obsolete methods?

Reply:
Restricting to obsolete methods sounds like a good idea.

Reply:
Can you confirm my observation above, that EMCP jmethodIDs are replaced?  I haven't looked at this code in a while.  Thanks.

Reply:
I am going to take a look.

Reply:
Ok, I found it.
The reason for the jmethodID not being cleaned out is this assignment of a new jmethodID to obsolete methods - https://github.com/openjdk/jdk/blob/a2c5f1fc914ef5c28d044b75598f895cf6097138/src/hotspot/share/prims/jvmtiRedefineClasses.cpp#L3887

Since this is not done for EMCP (or more generic, non-obsolete) methods restricting the cleanup process to obsolete methods still sounds like the right thing to do.

Reply:
Excellent thanks.